File: 0125-valid-palindrome/0125-valid-palindrome.java
class Solution {
    // pattern : 2 pointers 
    // time : O(N)
    // space : O(1)
    public boolean isPalindrome(String s) {
        int i = 0 ;
        int j = s.length() - 1; 
        while(i<=j){
            while(!Character.isLetterOrDigit(s.charAt(i)) && i<j)
                i++;
            while(!Character.isLetterOrDigit(s.charAt(j)) && i<j)
                j--;
            if (Character.toLowerCase(s.charAt(i)) != Character.toLowerCase(s.charAt(j))) 
                return false;
            i++;
            j--;
            
        }
        return true;
    }
}

File: 0125-valid-palindrome/0125-valid-palindrome.cpp
class Solution {
public:
    bool isPalindrome(string s) {
        return recursion(0, s.length() - 1, s);
    }

    bool recursion(int si, int ei, string &s) {
        if (ei - si < 1)
            return true;
        if (!isalnum(s[si]))
            return recursion(si + 1, ei, s);
        if (!isalnum(s[ei]))
            return recursion(si, ei - 1, s);
        if (tolower(s[si]) != tolower(s[ei]))
            return false;
        return recursion(si + 1, ei - 1, s);
    }
};
